plan.go

```diff
@@ -0,0 +1,206 @@
+/*
+ * go-leia
+ * Copyright (C) 2021 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package leia
+
+import (
+	"errors"
+
+	"go.etcd.io/bbolt"
+)
+
+// queryPlan is the interface for all query plans
+type queryPlan interface {
+	// execute the plan call the DocWalker for each matching document
+	execute(walker DocWalker) error
+}
+
+// defaultQueryPlan contains elements common for each query plan
+type defaultQueryPlan struct {
+	collection *collection
+}
+
+// fullTableScanQueryPlan is a query plan which scans all documents
+type fullTableScanQueryPlan struct {
+	defaultQueryPlan
+	queryParts []QueryPart
+}
+
+// resultScanQueryPlan is a query plan that uses an index and filters the results with the remaining query params
+type resultScanQueryPlan struct {
+	defaultQueryPlan
+	index Index
+	query Query
+}
+
+// indexScanQueryPlan is a special query plan that only loops over the index keys and document references
+type indexScanQueryPlan struct {
+	defaultQueryPlan
+	index Index
+	query Query
+}
+
+// ReferenceScanFn is a function type which is called with an index key and a document Reference as value
+type ReferenceScanFn func(key []byte, value []byte) error
+
+// documentScanFn is a function type which is called with a document Reference as key and a the document bytes as value
+type documentScanFn func(key []byte, value []byte) error
+
+func (f fullTableScanQueryPlan) execute(walker DocWalker) error {
+	return f.collection.db.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte(f.collection.Name))
+		if bucket == nil {
+			// no bucket means no docs
+			return nil
+		}
+		bucket = bucket.Bucket([]byte(documentCollection))
+		if bucket == nil {
+			// no bucket means no docs
+			return nil
```

Comment:
can this happen in normal circumstances, the collection bucket exists but the document bucket doesn't?

Reply:
yes when you create an index but don't add any documents. So each empty collection.